onnxruntime/core/providers/cpu/generator/range.cc

```diff
@@ -80,24 +81,43 @@ static Status ComputeRange(OpKernelContext* ctx) {
   return Status::OK();
 }
 
+namespace range_internal {
+
+// Workaround GCC bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47226
+// can not specify lambda directly
+template <class T, class... Args>
+inline int invoke_computerange_callable(size_t& called, Status& s, int32_t dt_type, Args&&... args) {
+  auto fn = [&] {
+    if (utils::ToTensorDataType<T>() == dt_type) {
+      s = ComputeRange<T>(std::forward<Args>(args)...);
+      ++called;
+    }
+    return 0;
+  };
+  return fn();
+}
+
+
+template<typename... Types>
+inline Status CallDispatcher(int32_t dt_type, OpKernelContext* ctx) {
+  Status s;
+  size_t called = 0;
+
+  int results[] = {0, invoke_computerange_callable<Types>(called, s, dt_type, ctx)...};
+
+  ORT_UNUSED_PARAMETER(results);
+  ORT_ENFORCE(called < 2, "Range CallDispatcher broken. Check for duplicate type.");
+  ORT_ENFORCE(called == 1, "Range op: Unsupported tensor data type:", dt_type);
```

Comment:
Is this possible? Doesn't it exit on the first match?

Reply:
No, it does not. To populate the array an instance of a function for all types will be called. Those that do not match the type will simply exit w/o running the functionality. However, if the coder makes a mistake and includes a duplicate type, then the same function might run twice.

---
In reply to: [341920577](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r341920577) [](ancestors = 341920577)